contracts/crypto-verify/src/contract.rs

```diff
@@ -0,0 +1,192 @@
+
+    const CREATOR: &str = "creator";
+    const SENDER: &str = "sender";
+
+    const MESSAGE_HEX: &str = "5c868fedb8026979ebd26f1ba07c27eedf4ff6d10443505a96ecaf21ba8c4f0937b3cd23ffdc3dd429d4cd1905fb8dbcceeff1350020e18b58d2ba70887baa3a9b783ad30d3fbf210331cdd7df8d77defa398cdacdfc2e359c7ba4cae46bb74401deb417f8b912a1aa966aeeba9c39c7dd22479ae2b30719dca2f2206c5eb4b7";
+    const SIGNATURE_HEX: &str = "207082eb2c3dfa0b454e0906051270ba4074ac93760ba9e7110cd9471475111151eb0dbbc9920e72146fb564f99d039802bf6ef2561446eb126ef364d21ee9c4";
+    const PUBLIC_KEY_HEX: &str = "04051c1ee2190ecfb174bfe4f90763f2b4ff7517b70a2aec1876ebcfd644c4633fb03f3cfbd94b1f376e34592d9d41ccaf640bb751b00a1fadeb0c01157769eb73";
+
+    fn setup() -> OwnedDeps<MockStorage, MockApi, MockQuerier> {
+        let mut deps = mock_dependencies(&[]);
+        let msg = InitMsg {};
+        let info = mock_info(CREATOR, &[]);
+        let res = init(deps.as_mut(), mock_env(), info, msg).unwrap();
+        assert_eq!(0, res.messages.len());
+        deps
+    }
+
+    #[test]
+    fn init_works() {
+        setup();
+    }
+
+    #[test]
+    fn verify_works() {
+        let mut deps = setup();
+
+        let message = hex::decode(MESSAGE_HEX).unwrap();
+        let signature = hex::decode(SIGNATURE_HEX).unwrap();
+        let public_key = hex::decode(PUBLIC_KEY_HEX).unwrap();
+
+        let verify_msg = HandleMsg::VerifySignature {
+            message: Binary(message),
+            signature: Binary(signature),
+            public_key: Binary(public_key),
+        };
+        let res = handle(
+            deps.as_mut(),
+            mock_env(),
+            mock_info(SENDER, &[]),
+            verify_msg,
+        )
+        .unwrap();
+        assert_eq!(
+            res,
+            Response {
+                messages: vec![],
+                attributes: vec![attr("action", "handle_verify")],
+                data: Some(Binary(vec![1]))
+            }
+        );
+    }
+
+    #[test]
+    fn verify_fails() {
+        let mut deps = setup();
+
+        let mut message = hex::decode(MESSAGE_HEX).unwrap();
+        // alter message
+        message[0] ^= 0x01;
+        let signature = hex::decode(SIGNATURE_HEX).unwrap();
+        let public_key = hex::decode(PUBLIC_KEY_HEX).unwrap();
+
+        let verify_msg = HandleMsg::VerifySignature {
+            message: Binary(message),
+            signature: Binary(signature),
+            public_key: Binary(public_key),
+        };
+        let res = handle(
+            deps.as_mut(),
+            mock_env(),
+            mock_info(SENDER, &[]),
+            verify_msg,
+        )
+        .unwrap();
+        assert_eq!(
+            res,
+            Response {
+                messages: vec![],
+                attributes: vec![attr("action", "handle_verify")],
+                data: Some(Binary(vec![0]))
+            }
+        );
+    }
+
+    #[test]
+    #[should_panic(expected = "empty")]
```

Comment:
I'm sure you have discussed this before. I am not a fan of panic in a reasonable error case in a contract, cuz you get no info out "contract failed - panic". If you can get some reasonable error message saying what the user did wrong, it is nice to be able to return that as an Error type.

Reply:
Yes, I have the same feeling. We discussed this with @webmaster128. Will revisit, to see what's best.

Reply:
I think the problem here is that those error messages come from VM land, and need to cross the boundary, which has some issues.

Reply:
Will merge this into the `crypto-api-package` branch, and we can continue the discussion there in any case.

Reply:
This particular test tests the behaviour of the mock implementations (which might not be exactly what we want to do). The production behaviour is a failure in the VM that never returns to the contract and is reported to the transaction log.

What both have in common:
1. The error case cannot be handled in the contract.
2. The error is reported (either by the unit tests or by the VM). This is not a panic created in the contract, as Ethan is referring to.

Reply:
There's now a [separate PR](#785) implementing error codes, for returning errors to contracts / clients.